// my-sql/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.escape = exports.escapeId = exports.format = exports.createPoolCluster = exports.createConnection = exports.createQuery = exports.Types = exports.createPool = exports.removeMiddleware = exports.setMiddleware = exports.addInternalPool = exports.endSession = exports.initSession = exports.executeTransaction = exports.executeQuery = undefined;

var _mysql = require('mysql');

Object.keys(_mysql).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    Object.defineProperty(exports, key, {
        enumerable: true,
        get: function get() {
            return _mysql[key];
        }
    });
});

var _mysql2 = _interopRequireDefault(_mysql);

var _mysqlFacade = require('./mysqlFacade');

var _mysqlFacade2 = _interopRequireDefault(_mysqlFacade);

var _dbManager = require('./dbManager');

var _dbManager2 = _interopRequireDefault(_dbManager);

var _builder = require('./query/builder');

var _builder2 = _interopRequireDefault(_builder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Created by pubudud on 8/29/17.
 */

Error.prototype.appendDetails = function (className, method, cause) {
    this.path = (this.path || '#') + ' -> [' + className + ']|(' + method + ')';
    this.causes = (this.causes || '#') + ' -> (' + method + ')|' + cause;
};

exports.default = _mysql2.default;
var executeQuery = exports.executeQuery = _mysqlFacade2.default.execute.bind(_mysqlFacade2.default);
var executeTransaction = exports.executeTransaction = _mysqlFacade2.default.executeTransaction.bind(_mysqlFacade2.default);

var initSession = exports.initSession = _dbManager2.default.init.bind(_dbManager2.default);
var endSession = exports.endSession = _dbManager2.default.end.bind(_dbManager2.default);

var addInternalPool = exports.addInternalPool = _dbManager2.default.createConnectionPool.bind(_dbManager2.default);
var setMiddleware = exports.setMiddleware = _builder2.default.use.bind(_builder2.default);
var removeMiddleware = exports.removeMiddleware = _builder2.default.removeMiddleware.bind(_builder2.default);

var createPool = exports.createPool = _mysql.createPool;
var Types = exports.Types = _mysql.Types;
var createQuery = exports.createQuery = _mysql.createQuery;
var createConnection = exports.createConnection = _mysql.createConnection;
var createPoolCluster = exports.createPoolCluster = _mysql.createPoolCluster;
var format = exports.format = _mysql.format;
var escapeId = exports.escapeId = _mysql.escapeId;
var escape = exports.escape = _mysql.escape;
