// my-sql/dist/__test__/mysqlStubManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * Created by pubudud on 9/4/17.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      */

var _sinon = require('sinon');

var _sinon2 = _interopRequireDefault(_sinon);

var _mysql = require('mysql');

var _mysql2 = _interopRequireDefault(_mysql);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Stub Manager for Device Dao
 */
var DeviceDaoStubManager = function () {

    /**
     * Constructor
     */
    function DeviceDaoStubManager() {
        _classCallCheck(this, DeviceDaoStubManager);

        this.sandbox = _sinon2.default.sandbox.create();
    }

    /**
     * Initialize all stubs of this module
     */


    _createClass(DeviceDaoStubManager, [{
        key: 'initStubs',
        value: function initStubs() {}
        // this.restoreStubs();
        // this.initCreatePool();


        /**
         * Restore all stubs of this module
         */

    }, {
        key: 'restoreStubs',
        value: function restoreStubs() {
            this.sandbox.restore();
        }

        /**
         * Init stub for getFirstDevice method
         * @returns {Object} - stub for reference
         */

    }, {
        key: 'initCreatePool',
        value: function initCreatePool() {
            var createPool = this.sandbox.stub(_mysql2.default, 'createPool');

            var getConnectionStub = this.sandbox.stub();

            getConnectionStub.returns({
                execute: _sinon2.default.spy()
            });

            createPool.returns({
                getConnection: getConnectionStub,
                end: _sinon2.default.spy()
            });

            return createPool;
        }
    }]);

    return DeviceDaoStubManager;
}();

exports.default = new DeviceDaoStubManager();
